### PR TITLE
feat: implement slider

### PR DESCRIPTION
**Pull request by TranQuangPhi**
## Features
- implement Slider and Slider range

## Check List
- [x] Your code builds clean without any errors or warnings.
- [x] Check coding convention.
- [x] Updated Storybook components and stories to reflect any changes made to the UI.
- [ ] Test covers all edge cases and coverage is greater than or equal to 80%.

## Proof of Completeness
IOS

https://user-images.githubusercontent.com/105619840/230307583-09085475-5915-480a-89db-09436fbd1aca.mp4

Android

https://user-images.githubusercontent.com/105619840/230307663-bbe0b97d-e8b2-4604-99d2-e8001e45e015.mov

### CODE REVIEW

File: src/components/Slider/Slider.tsx [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+  | {hasTrackPoint?: false; sliderWidth?: number}
+)
+
+type SliderProps = SliderCommonProps & SliderPropsWithOptionalWidth
+
+type SliderComponentProps = React.FunctionComponent<SliderProps> & {
+  Range: React.FunctionComponent<SliderProps>
+}
+
+const Slider: SliderComponentProps = ({
```
Comment (honghoangsts) [inline src/components/Slider/Slider.tsx:126]: const Slider : React.FC< SliderComponentProps > = () => ({})

File: src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+  minimumValue = 1,
+  maximumValue = 10,
+  step = DEFAULT_STEP,
+  trackStyle,
+  bgColorTrack = '#F1F1F1',
```
Comment (honghoangsts) [inline src/components/Slider/Slider.tsx:131]: Don't use magic string

File: src/components/Slider/Slider.tsx [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+type SliderProps = SliderCommonProps & SliderPropsWithOptionalWidth
+
+type SliderComponentProps = React.FunctionComponent<SliderProps> & {
+  Range: React.FunctionComponent<SliderProps>
+}
+
+const Slider: SliderComponentProps = ({
+  minimumValue = 1,
+  maximumValue = 10,
```
Comment (honghoangsts) [inline src/components/Slider/Slider.tsx:128]: Don't use magic number

File: src/components/Slider/Constants.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+const DEFAULT_STEP = 1
+const FIRST_POINT = 0
+const MINIMUM_TRACK_WIDTH = 0
+const defaultHitSlop = {top: 5, right: 5, bottom: 5, left: 5}
+
+const ThumbPosition = {left: 'left', right: 'right'}
+
+export {DEFAULT_STEP, FIRST_POINT, MINIMUM_TRACK_WIDTH, defaultHitSlop, ThumbPosition}
```
Comment (honghoangsts) [inline src/components/Slider/Constants.ts:8]: file name should be `constants.ts`

**Review by honghoangsts — CHANGES_REQUESTED** [review]
- Please remove all the magic number and magic string.
- Integrate with the theme to get default colors.
- Implement the way to let user can custom style from themes

File: src/components/Slider/Slider.tsx (hunk 2 above)
Comment (TranQuangPhi) [reply src/components/Slider/Slider.tsx]: done

File: src/components/Slider/Slider.tsx (hunk 3 above)
Comment (TranQuangPhi) [reply src/components/Slider/Slider.tsx]: done

File: src/components/Slider/Constants.ts (hunk 4 above)
Comment (TranQuangPhi) [reply src/components/Slider/Constants.ts]: done

File: src/components/Slider/Slider.tsx (hunk 1 above)
Comment (TranQuangPhi) [reply src/components/Slider/Slider.tsx:139]: I changed 'React.FunctionalComponent' to 'React.FC', but now I need to use '&' for the compound component

type SliderComponentProps = React.FC<SliderProps> & {
  Range: React.FC<ISliderRange>
}

File: src/components/Slider/Slider.tsx [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+  thumbSize = {width: metrics.medium, height: metrics.medium},
+  trackPointStyle,
+  onValueChange = () => null,
+}: SliderProps) => {
+  const sliderInfo = useSharedValue<SliderInfo>({
+    range: INIT_VALUE,
+    trackWidth: INIT_VALUE,
+  })
+  const sliderValue = useSharedValue<number>(INIT_POINT)
+  const progress = useSharedValue<number>(INIT_VALUE)
```
Comment (hangnguyensaigontech) [inline src/components/Slider/Slider.tsx:171]: Why we don't create just 1 useSharedValue for all related to slider?
Also how about change to **SliderState**  instead for correct meaning , for example:
```suggestion
  const sliderState = useSharedValue<SliderState>({
    range: INIT_VALUE,
    trackWidth: INIT_VALUE,
    currentValue: INIT_POINT,
    progress: INIT_VALUE
  })
```

File: src/components/Slider/Slider.tsx [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    sliderInfo.value = {range: range, trackWidth: width}
+  }
+
+  /**
+   * Function called when the user presses on a point on the slider's track
+   * Calculates the position of the pressed point on the slider's track and updates the slider's state accordingly
+   * @param {number} point - The index of the point on the slider's track that was pressed by the user
+   */
+  const onPressPoint = useCallback(
+    (point: number) => {
```
Comment (hangnguyensaigontech) [inline src/components/Slider/Slider.tsx:287]: recommend change to pointIndex or index for matching meaning. This named **point** sounds like it is the value not the index

File: src/components/Slider/Slider.tsx [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+      const positionPoint = sliderInfo.value.range * (point + FIRST_POINT)
+      const curPoint = point + FIRST_POINT
+      const value = minimumValue + curPoint * stepValue
```
Comment (hangnguyensaigontech) [inline src/components/Slider/Slider.tsx:290]:
```suggestion
      const curPoint = point + FIRST_POINT
      const positionPoint = sliderInfo.value.range * curPoint
      const value = minimumValue + curPoint * stepValue
```

File: src/components/Slider/Slider.tsx [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+      updateSlider(positionPoint, value)
+      onValueChange(value)
+    },
+    [minimumValue, onValueChange, sliderInfo.value.range, stepValue, updateSlider],
+  )
+
+  return (
+    <Container style={[!!sliderWidth && {width: sliderWidth}, style]}>
+      <Track style={trackStyle} onLayout={getTrackWidth} />
```
Comment (hangnguyensaigontech) [inline src/components/Slider/Slider.tsx:300]: Wrapper all component with <Track></Track> instead of <Container></Container>
Style from **<Container>** can be move to **Track** and it will work correctly, no need to wrapper another View just to set width. it's redundant

File: src/components/Slider/components/Thumb.tsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  <PanGestureHandler onGestureEvent={onGestureEvent}>
+    <ThumbContainer
+      thumbSize={thumbSize}
+      hasThumbComponent={!!thumbComponent}
+      style={[thumbStyle, animatedThumbStyle]}>
+      <LabelContainer
+        background={bgColorLabelView}
+        style={!alwaysShowValue && opacityStyle}
+        thumbSize={thumbSize}>
+        <TriangleDown background={bgColorLabelView} style={{transform: [{rotate: '180deg'}]}} />
```
Comment (hangnguyensaigontech) [inline src/components/Slider/components/Thumb.tsx:53]: style={{transform: [{rotate: '180deg'}]}}

why used inline style and not move in here instead

image

File: src/components/Slider/components/Thumb.tsx [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  borderRightWidth: 5,
+  borderBottomWidth: 10,
+  borderLeftColor: 'transparent',
+  borderRightColor: 'transparent',
+  borderBottomColor: background || theme?.colors.primary,
+}))
+
+const LabelContainer = styled(Animated.View)((props: ThumbContainerStyle) => ({
+  position: 'absolute' as Position,
+  top: -(props.theme?.spacing.xxxl ?? 40),
```
Comment (hangnguyensaigontech) [inline src/components/Slider/components/Thumb.tsx:88]: not use number, use from metrics instead

File: src/components/Slider/components/Thumb.tsx [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+const TriangleDown = styled.View(({background, theme}: {background?: string; theme: ITheme}) => ({
+  position: 'absolute' as Position,
+  bottom: -5,
+  width: 0,
+  height: 0,
+  backgroundColor: 'transparent',
+  borderStyle: 'solid',
+  borderLeftWidth: 5,
+  borderRightWidth: 5,
+  borderBottomWidth: 10,
```
Comment (hangnguyensaigontech) [inline src/components/Slider/components/Thumb.tsx:80]: use defined value from metrics

File: src/components/Slider/SliderFixedRange.tsx [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+
+type ContainerProps = {
+  width: number
+}
+
+type NumberRange = {
+  maximum: number
+  minimum: number
+}
+
```
Comment (hangnguyensaigontech) [inline src/components/Slider/SliderFixedRange.tsx:52]: export reuse in component SliderRange, not duplicate type again

File: src/components/Slider/SliderFixedRange.tsx [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+  }))
+
+  /**
+   * Display the current slider value in real time as the user slides the thumb along the slider track
+   */
+  const leftAnimatedProps = useAnimatedProps(
+    () =>
+      ({
+        text: `${minimumValue + sliderValue.value.left * step}`,
+      } as AnimatedLabelProps),
```
Comment (ledutu2) [inline src/components/Slider/SliderFixedRange.tsx:238]: I think the arrow function should be on the same line with useAnimatedProps

File: src/components/Slider/SliderFixedRange.tsx [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+  )
+
+  const rightAnimatedProps = useAnimatedProps(
+    () =>
+      ({
+        text: `${minimumValue + sliderValue.value.right * step}`,
+      } as AnimatedLabelProps),
+  )
```
Comment (ledutu2) [inline src/components/Slider/SliderFixedRange.tsx:246]: same above

File: src/components/Slider/Slider.tsx (hunk 5 above)
Comment (anhle10051996) [reply src/components/Slider/Slider.tsx:172]: sliderState and sliderValue + progress is using for different purpose. So i think it more clearly to keep it separate like above

File: src/components/Slider/Slider.tsx [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+        text: `${roundToValue !== undefined ? sliderValue.value.toFixed(roundToValue) : sliderValue.value}`,
+      } as AnimatedLabelProps),
+  )
+
+  const getTrackWidth = (event: LayoutChangeEvent) => {
+    const {width} = event.nativeEvent.layout
+    // Range refers to the width of a point
+    // It is used to calculate the correct position of the slider while sliding
+    const range = width / totalPoint
+    sliderInfo.value = {range: range, trackWidth: width}
```
[the remark this answers is not in the document]
Comment (anhle10051996) [reply src/components/Slider/Slider.tsx:279]: updated

File: src/components/Slider/Slider.tsx (hunk 6 above)
Comment (anhle10051996) [reply src/components/Slider/Slider.tsx:288]: updated

File: src/components/Slider/Slider.tsx (hunk 7 above)
Comment (anhle10051996) [reply src/components/Slider/Slider.tsx:291]: updated

File: src/components/Slider/SliderFixedRange.tsx (hunk 13 above)
Comment (anhle10051996) [reply src/components/Slider/SliderFixedRange.tsx:238]: the line should maximum should be limited to of 80 characters

File: src/components/Slider/SliderFixedRange.tsx (hunk 14 above)
Comment (anhle10051996) [reply src/components/Slider/SliderFixedRange.tsx:246]: the line should maximum should be limited to of 80 characters

File: src/components/Slider/components/Thumb.tsx (hunk 11 above)
Comment (anhle10051996) [reply src/components/Slider/components/Thumb.tsx:80]: i think it's not necessary. the width in different case have different and we cannot define all the number in theme

File: src/components/Slider/SliderFixedRange.tsx (hunk 12 above)
Comment (anhle10051996) [reply src/components/Slider/SliderFixedRange.tsx:52]: updated

File: src/components/Slider/components/Thumb.tsx (hunk 10 above)
Comment (anhle10051996) [reply src/components/Slider/components/Thumb.tsx]: updated

File: src/components/Slider/Slider.tsx (hunk 8 above)
Comment (anhle10051996) [reply src/components/Slider/Slider.tsx:301]: The Container will style for both <Track /> and <Tracked /> component. moving style to Track will make the component work incorrectly

File: src/components/Slider/SliderFixed.tsx [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+  thumbComponent,
+  showTrackPoint,
+  tapToSeek,
+  hitSlopPoint = hitSlop,
+  sliderWidth,
+  thumbSize = {width: metrics.medium, height: metrics.medium},
+  trackPointStyle,
+  onValueChange = () => null,
+}: SliderProps) => {
+  const sliderInfo = useSharedValue<SliderInfo>({range: INIT_VALUE, trackWidth: INIT_VALUE})
```
[the remark this answers is not in the document]
Comment (anhle10051996) [reply src/components/Slider/SliderFixed.tsx:52]: The Container will style for both and component. moving style to Track will make the component work incorrectly

File: src/components/Slider/components/Thumb.tsx (hunk 9 above)
Comment (anhle10051996) [reply src/components/Slider/components/Thumb.tsx]: updated

**Review by ngochuyduong — APPROVED** [review]

File: src/components/Slider/SliderFixed.tsx [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+  labelStyle,
+  thumbComponent,
+  showTrackPoint,
+  tapToSeek,
+  hitSlopPoint = hitSlop,
+  sliderWidth,
+  thumbSize = {width: metrics.medium, height: metrics.medium},
+  trackPointStyle,
+  onValueChange = () => null,
+}: SliderProps) => {
```
Comment (baonguyenhsts) [inline src/components/Slider/SliderFixed.tsx:52]: Please remove the `SliderProps` because it is unnecessary

File: src/components/Slider/constants.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+const DEFAULT_MAXIMUM_VALUE = 1
+const INIT_VALUE = 0
+const VISIBLE = 1
+const INVISIBLE = 0
+const NEXT_STEP = 1
+const PREVIOUS_STEP = 1
+const FIRST_POINT = 1
+const DURATION = 200
+
+const ThumbPosition = {left: 'left', right: 'right'}
```
Comment (baonguyenhsts) [inline src/components/Slider/constants.ts:14]: Please change `ThumbPosition` to `THUMB_POSITION` and check the relevant files with this const

**Review by baonguyenhsts — CHANGES_REQUESTED** [review]

**Review by loc-nguyenthien — APPROVED** [review]

**Review by baonguyenhsts — APPROVED** [review]

**Review by vunguyendhSTS — CHANGES_REQUESTED** [review]

**Review by vunguyendhSTS — APPROVED** [review]

**Review by honghoangsts — APPROVED** [review]
